gdx/src/com/badlogic/gdx/scenes/scene2d/ui/Spinner.java

```diff
@@ -0,0 +1,262 @@
+				if (getStage() != null) {
+					getStage().setScrollFocus(Spinner.this);
+					getStage().setKeyboardFocus(Spinner.this);
+				}
+				updateBackgroundForCoords(x, y);
+			}
+
+			@Override
+			public void exit (InputEvent event, float x, float y, int pointer, Actor toActor) {
+				updateBackgroundForCoords(x, y);
+			}
+
+			@Override
+			public boolean scrolled (InputEvent event, float x, float y, int amount) {
+				// Update spinner value if mouse is hovering over widget
+				if (getState() != SpinnerState.IDLE) updateValue(amount < 0);
+				return false;
+			}
+
+			@Override
+			public boolean keyDown (InputEvent event, int keycode) {
+				// Update spinner value if mouse is hovering over widget
+				if (getState() != SpinnerState.IDLE) {
+					if (keycode == Input.Keys.UP)
+						updateValue(true);
+					else if (keycode == Input.Keys.DOWN) updateValue(false);
+				}
+				return false;
+			}
+
+		});
+	}
+
+	/** This method is called by any event that triggers a modification to the spinner's current value.
+	 * @param increase True for spinner's next value and false for spinner's previous value */
+	protected void updateValue (boolean increase) {
+		// Update value, and if change occurred, dispatch a change event
+		if ((increase && getModel().next()) || (!increase && getModel().previous())) {
+			layout.setText(style.font, model.getDisplayValue()); // Update layout for new current value
+			ChangeEvent changeEvent = Pools.obtain(ChangeEvent.class);
+			fire(changeEvent);
+			Pools.free(changeEvent);
+		}
+		invalidate();
+	}
+
+	protected SpinnerState getState () {
+		return state;
+	}
+
+	protected void setState (SpinnerState state) {
+		this.state = state;
+	}
+
+	/** Updates the spinner's background and state for the given coordinates
+	 * @param x Current x coordinate of mouse
+	 * @param y Y Current y coordinate of mouse */
+	protected void updateBackgroundForCoords (float x, float y) {
+		float diffX = getWidth() - x;
+		float diffY = y;
+		state = SpinnerState.IDLE;
+		currentBackground = getStyle().background;
+		if (Gdx.input.isButtonPressed(Buttons.LEFT) && diffX < getStyle().background.getRightWidth() && diffX >= 0) {
+			if (diffY <= getHeight() && diffY > getHeight() / 2) {
+				state = SpinnerState.NEXT;
+				currentBackground = getStyle().backgroundNext;
+			} else if (diffY >= 0 && diffY <= getHeight() / 2) {
+				state = SpinnerState.PREVIOUS;
+				currentBackground = getStyle().backgroundPrev;
+			}
+		} else if (diffX >= 0 && diffX <= getWidth() && diffY >= 0 && diffY <= getHeight()) {
+			state = SpinnerState.HOVER;
+			currentBackground = getStyle().backgroundHover;
+		}
+		invalidate();
+	}
+
+	@Override
+	public void draw (Batch batch, float parentAlpha) {
+		currentBackground.draw(batch, getX(), getY(), getWidth(), getHeight());
+		// Center the text
+		style.font.draw(batch, model.getDisplayValue() + "", getX() + style.background.getLeftWidth(),
```

Comment:
You already set the display value text into the `layout`, you can (and should) also use it for drawing. Also, this allocates garbage even if `getDisplayValue` returns `String`.